nexus/src/db/queries/network_interface.rs

```diff
@@ -979,16 +989,342 @@ impl QueryFragment<Pg> for InsertNetworkInterfaceQueryValues {
+        NETWORK_INTERFACE_FROM_CLAUSE.walk_ast(out.reborrow())?;
+        out.push_sql(" WHERE ");
+        out.push_identifier(dsl::instance_id::NAME)?;
+        out.push_sql(" = ");
+        out.push_bind_param::<sql_types::Uuid, Uuid>(&self.instance_id)?;
+        out.push_sql(" AND ");
+        out.push_identifier(dsl::time_deleted::NAME)?;
+        out.push_sql(" IS NULL LIMIT 1)");
+        Ok(())
+    }
+}
+
+type InstanceFromClause = FromClause<db::schema::instance::table>;
+const INSTANCE_FROM_CLAUSE: InstanceFromClause = InstanceFromClause::new();
+
+/// Delete a network interface from an instance.
+///
+/// There are a few preconditions that need to be checked when deleting a NIC.
+/// First, the instance must currently be stopped, though we may relax this in
+/// the future. Second, while an instance may have zero or more interfaces, if
+/// it has one or more, exactly one of those must be the primary interface. That
+/// means we can only delete the primary interface if there are no secondary
+/// interfaces. The full query is:
+///
+/// ```sql
+/// WITH
+///     interface AS MATERIALIZED (
+///         SELECT CAST(IF(
+///             (
+///                 SELECT
+///                     NOT is_primary
+///                 FROM
+///                     network_interface
+///                 WHERE
+///                     id = <interface_id> AND
+///                     time_deleted IS NULL
+///             )
+///                 OR
+///             (
+///                 SELECT
+///                     COUNT(1)
+///                 FROM
+///                     network_interface
+///                 WHERE
+///                     instance_id = <instance_id> AND
+///                     name = <name>
+///                     time_deleted IS NULL
+///             ) = 1,
+///             '<interface_id>',
+///             'has-secondary'
+///         ) AS UUID)
+///     )
+///     instance AS MATERIALIZED (
+///         SELECT CAST(IF(
+///             EXISTS(
+///                 SELECT
+///                     id
+///                 FROM
+///                     instance
+///                 WHERE
+///                     id = <instance_id> AND
+///                     time_deleted IS NULL AND
+///                     state = 'stopped'
+///             ),
+///             '<instance_id>',
+///             'inst-running'
+///         ) AS UUID)
+///     )
+/// UPDATE
+///     network_interface
+/// SET
+///     time_deleted = NOW()
+/// WHERE
+///     id = <interface_id> AND
+///     time_deleted IS NULL
+/// ```
+#[derive(Debug, Clone)]
+pub struct DeleteNetworkInterfaceQuery {
+    interface_id: Uuid,
+    instance_id: Uuid,
+    instance_id_str: String,
+    instance_state: db::model::InstanceState,
+}
+
+impl DeleteNetworkInterfaceQuery {
+    pub fn new(instance_id: Uuid, interface_id: Uuid) -> Self {
+        Self {
+            interface_id,
+            instance_id,
+            instance_id_str: instance_id.to_string(),
```

Comment:
See comment above - what's the point of passing this as a name, if it's always implied by the ID?

Reply:
I think I understand your question, and this could be better commented. The point of passing the instance ID as both a UUID and a string is to generate a particular type of error that enables us to recover why the query failed.

This query, and a few others in this module, use some casting trickery to do validation. That generally looks like:

```sql
CAST(IF(<data is valid>, 'uuid-as-a-string', 'garbage') AS UUID)
```

So here, the check we're doing is that this interface is a secondary, or it is the only primary. If that succeeds we cast the UUID as a string back to a UUID. If that fails, we try to cast some junk string to a UUID. That failure is included in the database error message, so we can determine why the query failed just by looking at that message.

Reply:
Gotcha, this makes sense. If we're looking up interfaces by "just UUID" and using the name only for this casting, that seems totally reasonable.